pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -243,6 +411,210 @@ func (ctrl *PersistentVolumeController) isFullySynced() bool {
+		obj.Annotations = make(map[string]string)
+	}
+	obj.Annotations[ann] = "yes"
+}
+
+func (ctrl *PersistentVolumeController) updateClaimPhase(claim *api.PersistentVolumeClaim, phase api.PersistentVolumeClaimPhase) (*api.PersistentVolumeClaim, error) {
+	glog.V(5).Infof("Updating PersistentVolumeClaim[%s]: set phase %s", claimToClaimKey(claim), phase)
+	if claim.Status.Phase == phase {
+		// Nothing to do.
+		glog.V(5).Infof("Updating PersistentVolumeClaim[%s]: phase %s already set", claimToClaimKey(claim), phase)
+		return claim, nil
+	}
+
+	clone, err := conversion.NewCloner().DeepCopy(claim)
+	if err != nil {
+		return nil, fmt.Errorf("Error cloning claim: %v", err)
+	}
+	claimClone, ok := clone.(*api.PersistentVolumeClaim)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected claim cast error : %v", claimClone)
+	}
+
+	claimClone.Status.Phase = phase
+	newClaim, err := ctrl.kubeClient.Core().PersistentVolumeClaims(claimClone.Namespace).UpdateStatus(claimClone)
+	if err != nil {
+		glog.V(5).Infof("Updating PersistentVolumeClaim[%s]: set phase %s failed: %v", claimToClaimKey(claim), phase, err)
+		return newClaim, err
+	}
+	glog.V(5).Infof("Updating PersistentVolumeClaim[%s]: phase %s set", claimToClaimKey(claim), phase)
+	return newClaim, nil
+}
+
+func (ctrl *PersistentVolumeController) updateVolumePhase(volume *api.PersistentVolume, phase api.PersistentVolumePhase) (*api.PersistentVolume, error) {
+	glog.V(5).Infof("Updating PersistentVolume[%s]: set phase %s", volume.Name, phase)
+	if volume.Status.Phase == phase {
+		// Nothing to do.
+		glog.V(5).Infof("Updating PersistentVolume[%s]: phase %s already set", volume.Name, phase)
+		return volume, nil
+	}
+
+	clone, err := conversion.NewCloner().DeepCopy(volume)
+	if err != nil {
+		return nil, fmt.Errorf("Error cloning claim: %v", err)
+	}
+	volumeClone, ok := clone.(*api.PersistentVolume)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected volume cast error : %v", volumeClone)
+	}
+
+	volumeClone.Status.Phase = phase
+	newVol, err := ctrl.kubeClient.Core().PersistentVolumes().UpdateStatus(volumeClone)
+	if err != nil {
+		glog.V(5).Infof("Updating PersistentVolume[%s]: set phase %s failed: %v", volume.Name, phase, err)
+		return newVol, err
+	}
+	glog.V(5).Infof("Updating PersistentVolume[%s]: phase %s set", volume.Name, phase)
+	return newVol, err
+}
+
+func (ctrl *PersistentVolumeController) bindVolumeToClaim(volume *api.PersistentVolume, claim *api.PersistentVolumeClaim) (*api.PersistentVolume, error) {
+	glog.V(5).Infof("Updating PersistentVolume[%s]: binding to %q", volume.Name, claimToClaimKey(claim))
+
+	// Check if the volume was already bound
+	boundByController := false
+	if !isVolumeBoundToClaim(volume, claim) {
+		boundByController = true
+	}
+
+	// Save the PV only when any modification is neccesary.
+	if volume.Spec.ClaimRef == nil || volume.Spec.ClaimRef.Namespace != claim.Namespace ||
+		volume.Spec.ClaimRef.Name != claim.Name ||
+		volume.Spec.ClaimRef.UID != claim.UID ||
```

Comment:
isVolumeBoundToClaim() allows for ClaimRef.UID to be empty, "".  Why is a "" UID not tested for here?